tahowallet/extension
# Implement EIP-7524 (PLUME Signatures)

## Explanation
We want Taho to be the first wallet to support private voting and private airdrops on Ethereum and other EVM chains.

This PR adds new `eth_getPlumeSignature` RPC method that implements a [novel ECDSA nullifier scheme](https://aayushg.com/thesis.pdf) as [described](https://blog.aayushg.com/posts/nullifier) in [EIP-7524](https://github.com/ethereum/EIPs/blob/53683e7716ff7c962e5a30a087ced51a4c60951b/EIPS/eip-7524.md).

The `eth_getPlumeSignature` method takes in two parameters, a **message** and an **address**, then generates a deterministic signature (PLUME) and several other inputs. The plume can be used as a nullifier to prevent double-spending in an anonymity set. This capability unlocks novel on-chain behavior, such as [private DAO voting](https://prop.house/nouns/private-voting-research-sprint), [fair, non-doxxing airdrops](https://github.com/stealthdrop/stealthdrop), and more.

## Screenshot
<img width="496" alt="plume confirmation window" src="https://github.com/tahowallet/extension/assets/168856/4bd555e2-a873-42d1-b138-39e77d7b3dc7">


## Manual Testing Steps

After building and running Taho locally, enter this into the browser console

```js
await window.ethereum.request({
  "method": "eth_requestAccounts",
  "params": []
});

accountAddress = (await window.ethereum.request({
  "method": "eth_accounts",
  "params": []
}))[0];

await window.ethereum.request({
  "method": "eth_getPlumeSignature",
  "params": [
    "this is a test message - hi aayush",
    accountAddress
  ]
});
```

A confirmation screen should open up. After clicking "Sign", you will see the plume and other signals outputted into the console.
